net/procfs/net_tcp.c

```diff
@@ -0,0 +1,172 @@
+
+#include <nuttx/net/netstats.h>
+
+#include "procfs/procfs.h"
+#include "tcp/tcp.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#if defined(CONFIG_NET_TCP) && defined(CONFIG_NET_IPv4)
+
+#define TCP_LINELEN 120
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: netprocfs_read_tcpstats
+ *
+ * Description:
+ *   Read and format TCP statistics.
+ *
+ * Input Parameters:
+ *   priv - A reference to the network procfs file structure
+ *   buffer - The user-provided buffer into which network status will be
+ *            returned.
+ *   bulen  - The size in bytes of the user provided buffer.
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success; a negated errno value is returned
+ *   on failure.
+ *
+ ****************************************************************************/
+
+ssize_t netprocfs_read_tcpstats(FAR struct netprocfs_file_s *priv,
+                                FAR char *buffer, size_t buflen)
+{
+  FAR struct tcp_conn_s *conn = NULL;
+  char remote[INET_ADDRSTRLEN + 1];
+  char local[INET_ADDRSTRLEN + 1];
+  struct in_addr laddr;
+  struct in_addr raddr;
+  int skip = 1;
+  int len = 0;
+
+  net_lock();
+
+  if (tcp_nextconn(NULL) == NULL)
+    {
+      goto end_lock;
+    }
+
+  if (priv->offset == 0)
+    {
+      len = snprintf(buffer, buflen, "TCP sl     "
+                                     "local_address          "
+                                     "remote_address     "
+                                     "st flg ref tmr uack nrt   "
+#if CONFIG_NET_SEND_BUFSIZE > 0
+                                     "txsz   "
+#endif
+                                     "rxsz"
+#ifdef CONFIG_NET_IPv6
+                                     "  (IPv4)"
+#endif /* CONFIG_NET_IPv6 */
+                                     "\n");
+      priv->offset = 1;
+    }
+
+  local[INET_ADDRSTRLEN] = '\0';
+  remote[INET_ADDRSTRLEN] = '\0';
+
+  while ((conn = tcp_nextconn(conn)) != NULL)
+    {
+#if defined(CONFIG_NET_IPv4) && defined(CONFIG_NET_IPv6)
+      if (conn->domain != PF_INET)
+        {
+          continue;
+        }
+#endif /* CONFIG_NET_IPv4 && CONFIG_NET_IPv6 */
+
+      if (++skip <= priv->offset)
+        {
+          continue;
+        }
+
+      if (buflen - len < TCP_LINELEN)
+        {
+          break;
+        }
+
+      laddr.s_addr = conn->u.ipv4.laddr;
+      raddr.s_addr = conn->u.ipv4.raddr;
+
+      len += snprintf(buffer + len, buflen - len,
+                      " %2" PRIu8 ": %16s:%6" PRIu16 " %16s:%6" PRIu16,
+                      priv->offset++,
+                      inet_ntoa_r(laddr, local, INET_ADDRSTRLEN),
```

Comment:
let's call inet_ntop to support both ipv4 and ipv6

Reply:
Done